Arrays

VeraSimon opened the pull request:
## MVP Day 1
- [x] Create
- [x] Destroy
- [x] Read
- [x] Append

## MVP Day 2
- [ ] Insert (implemented, but having segfault issues with strdup)
- [x] Delete
- [x] Resize

VeraSimon committed:
gitignore update

VeraSimon committed:
Arrays
Implemented create_array and destroy_array. Nothing seems broken yet.

VeraSimon committed:
gitignore
Forgot to add *.exe

VeraSimon committed:
Arrays
Implemented arr_read. Updated create_array to malloc the correct size, instead of chars.

VeraSimon committed:
Arrays
Implemented arr_append. Need to circle back and address error conditions using errno.h.

VeraSimon committed:
Cleanup
Removed accidentally added test executable

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -19,45 +19,51 @@ typedef struct Array {
 
-  // Allocate memory for elements
+    // Allocate memory for elements
+    newArr->elements = malloc(capacity * sizeof(char));
```
`sizeof(char *)` :) remember this is an array of `char *`, not an array of chars

VeraSimon replied:
Outdated commit ;P

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -19,45 +19,51 @@ typedef struct Array {
 
-  // Free array
+    // Free all elements
+    // Only one of the elements is a pointer to something we manually allocated memory for, so here it is
```
veeeery true! unless of course you decide to implement `insert` and `append` like we did during lecture/in the solution code.
lecture's/solution code's `insert` and `append` methods don't point directly to the element added, but instead use `strdup()` to make a copy of that element being added. since `strdup()` malloc's under the hood, you'd need to free those as well.

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -29,7 +29,7 @@ Array *create_array(int capacity)
 
     // Allocate memory for elements
-    newArr->elements = malloc(capacity * sizeof(char));
+    newArr->elements = malloc(capacity * sizeof(newArr->elements));
```
this works because `newArr->elements` is a pointer (itself pointing to an array of (char *) ), so `sizeof()` still returns a pointer size. however, for humans reading this, `sizeof(char *)` better explains what you're trying to do

VeraSimon replied:
lol, quoting me at myself

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -79,8 +79,16 @@ char *arr_read(Array *arr, int index)
+    else
+    {
+        // Otherwise, return the element at the given index
+        return arr->elements[index];
```
perfect

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -116,10 +116,18 @@ void arr_append(Array *arr, char *element)
 
     // Resize the array if the number of elements is over capacity
     // or throw an error if resize isn't implemented yet.
+    if (arr->count >= arr->capacity)
```
👍

VeraSimon committed:
Arrays
Some minor tweaks based on todays lecture

VeraSimon committed:
Arrays
Implemented resize_array, making reading the actual arrays.c file easier at a glance.

VeraSimon committed:
Arrays
Implemented arr_insert

VeraSimon committed:
Arrays
Implemented arr_remove. Having issues with seg faults when using strdup in arr_insert. Will have to investigate further.

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -29,7 +29,7 @@ Array *create_array(int capacity)
 
     // Allocate memory for elements
-    newArr->elements = malloc(capacity * sizeof(newArr->elements));
+    newArr->elements = malloc(capacity * sizeof(char *));
```
👍

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -79,10 +84,10 @@ char *arr_read(Array *arr, int index)
     {
         fprintf(stderr, "Index %i out of range\n", index);
-        return -1;
+        return NULL;
```
:) I like to look at the tests from the very start. they give great hints about what's expected from your functions

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -7,7 +7,7 @@ typedef struct Array
     int capacity;    // How many elements can this array hold?
     int count;       // How many states does the array currently hold?
-    char **elements; // The string elements contained in the array (it's an array of strings!)
+    char **elements; // The string elements contained in the array (it's an array of string pointers!)
```
yeah! they just happen to be so similar in C, but it's good to separate the concepts

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -59,14 +59,21 @@ void destroy_array(Array *arr)
 {
-
     // Create a new element storage with double capacity
+    char **newElementArr = malloc(2 * sizeof(arr));
```
this is actually quite clever. nice!

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -59,14 +59,21 @@ void destroy_array(Array *arr)
+    free(arr->elements);
 
     // Update the elements and capacity to new values
+    arr->capacity *= 2;
```
also very clever!

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -123,9 +130,9 @@ void arr_append(Array *arr, char *element)
-        // TODO: Implement resize_array, then update this to resize the array.
-        fprintf(stderr, "Can't add element %s. Array is full.\n", element);
-        return;
+        resize_array(arr);
```
:)

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -109,15 +110,31 @@ char *arr_read(Array *arr, int index)
 
-    // Throw an error if the index is greater than the current count
+    // Throw an error if the index is greater than (or equal to) the current count
+    if (index >= arr->count)
```
what if `index == arr->count`?

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -109,15 +110,31 @@ char *arr_read(Array *arr, int index)
+    }
 
     // Move every element after the insert index to the right one position
+    for (int i = arr->count; i > index; i--)
```
from right to left, very nice

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -59,8 +70,10 @@ void destroy_array(Array *arr)
+
     // Create a new element storage with double capacity
-    char **newElementArr = malloc(2 * sizeof(arr));
+    char **newElementArr = calloc(2, sizeof(arr));
```
in case you miss it because I wrote it in the last commit: very clever you!

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -109,9 +139,10 @@ char *arr_read(Array *arr, int index)
-    // Throw an error if the index is greater than (or equal to) the current count
-    if (index >= arr->count)
+    // Throw an error if the index is greater than the current count
+    if (index >= arr->count && arr->count > 0)
```
I love how you think of _all_ edge cases

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -169,13 +222,45 @@ void arr_append(Array *arr, char *element)
+        }
+
+        // In case the passed element isn't in the array...
+        if (&arr->elements[index] == &arr->elements[lastElement] && arr->elements[index] != arr->elements[lastElement])
```
jaja nice, but would simply `index == lastElement` not work for the first half of your if statement?

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -169,13 +222,45 @@ void arr_append(Array *arr, char *element)
+    int lastElement = arr->count - 1;
+    // Saving for after the search loop so we know when to stop shifting elements down
+    int index = 0;
+    for (; index < arr->count; index++)
```
nothing here for the first part of the loop? didn't know you could do that!

LuisanSuarez commented on arrays/arrays.c:
```diff
@@ -169,13 +222,45 @@ void arr_append(Array *arr, char *element)
+    {
+        if (*arr->elements[index] == *element)
+        {
+            free(arr->elements[index]);
```
nice idea taking care of it right away
